File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    
    let count = 0;
    let map = new Map();
    const n = nums.length;
    let ps = new Array(n);
    for(let i = 0; i < n; i++) {
        ps[i] = (i > 0 ? ps[i-1] : 0) + nums[i];
    }
    map.set(0,1);
    for(let i = 0; i < n; i++) {
        if(map.has(ps[i] - k)) {
            count += map.get(ps[i] - k);
        }
        map.set(ps[i], (map.get(ps[i]) || 0) + 1);
    }
    return count;
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        final int n = nums.length;
        int sum = 0;
        int count = 0;
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        map.put(0,1);
        for(int i = 0; i < n; i++) {
            sum += nums[i];
            if(map.containsKey(sum - k)) {
                count += map.get(sum-k);
            }
            map.put(sum, map.getOrDefault(sum,0) + 1);
        }
        return count;
    }
}